pybricks/util_pb/pb_flash.c

```diff
@@ -0,0 +1,450 @@
+}
+
+static HAL_StatusTypeDef flash_block_erase(uint32_t address) {
+
+    address += 1024 * 1024;
+
+    // Enable write mode
+    HAL_StatusTypeDef err = write_enable();
+    if (err != HAL_OK) {
+        return err;
+    }
+
+    uint8_t address_bytes[5];
+    address_bytes[0] = 0x21;
+    address_bytes[1] = (address & 0xff000000) >> 24;
+    address_bytes[2] = (address & 0x00ff0000) >> 16;
+    address_bytes[3] = (address & 0x0000ff00) >> 8;
+    address_bytes[4] = address & 0x000000ff;
+
+    // Send erase command
+    flash_enable(true);
+    err = HAL_SPI_Transmit(&hspi2, address_bytes, sizeof(address_bytes), 100);
+    flash_enable(false);
+
+    err = flash_wait_ready();
+
+    return err;
+}
+
+int block_device_read(const struct lfs_config *c, lfs_block_t block, lfs_off_t off, void *buffer, lfs_size_t size) {
+
+    lfs_size_t done = 0;
+
+    while (done < size) {
+
+        // How many bytes to read in one go.
+        lfs_size_t read_now = size - done;
+        if (read_now > c->read_size) {
+            read_now = c->read_size;
+        }
+
+        // Read chunk of flash.
+        if (flash_read(block * c->block_size + off + done, buffer + done, read_now) != HAL_OK) {
+            return LFS_ERR_IO;
+        }
+        done += read_now;
+
+        // Give MicroPython and PBIO some time.
+        MICROPY_EVENT_POLL_HOOK;
+    }
+    return 0;
+}
+
+int block_device_prog(const struct lfs_config *c, lfs_block_t block, lfs_off_t off, const void *buffer, lfs_size_t size) {
+    lfs_size_t done = 0;
+
+    while (done < size) {
+
+        // How many bytes to write in one go.
+        lfs_size_t write_now = size - done;
+
+        // Must not be larger than write size
+        if (write_now > c->prog_size) {
+            write_now = c->prog_size;
+        }
+
+        // Must not wrap around the page
+        lfs_size_t write_max = 256 - ((off + done) & 0xFF);
+        if (write_now > write_max) {
+            write_now = write_max;
+        }
+
+        // Write chunk of flash.
+        if (flash_write(block * c->block_size + off + done, buffer + done, write_now) != HAL_OK) {
+            return LFS_ERR_IO;
+        }
+        done += write_now;
+
+        // Give MicroPython and PBIO some time.
+        MICROPY_EVENT_POLL_HOOK;
+    }
+
+    return 0;
+}
+
+int block_device_erase(const struct lfs_config *c, lfs_block_t block) {
+    // Erase block of flash. Block size assumed to match erase size.
+
+    if (flash_block_erase(block * c->block_size) != HAL_OK) {
+        return LFS_ERR_IO;
+    }
+    return LFS_ERR_OK;
+}
+
+int block_device_sync(const struct lfs_config *c) {
+    return 0;
+}
+
+lfs_t lfs;
+lfs_file_t file;
+
+uint8_t lfs_read_buf[256];
+uint8_t lfs_prog_buf[256];
+uint8_t lfs_lookahead_buf[256];
+uint8_t lfs_file_buf[256];
+
+struct lfs_config cfg = {
+    .read = block_device_read,
+    .prog = block_device_prog,
+    .erase = block_device_erase,
+    .sync = block_device_sync,
+
+    .read_size = 256,
```

Comment:
Does this info come from the data sheet? If so, we should have a comment explaining which chip we have.

Reply:
No. I think we can read an arbitrary lengths from the device.  Since reads are blocking for now, this provides a reasonable balance between read time and giving the run loop enough cycles.